File: android/app/src/main/java/ru/bgtu_voenmeh/zapara/data/api/TimetableApiClient.kt
package ru.bgtu_voenmeh.zapara.data.api

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import kotlinx.coroutines.withTimeout
import java.net.InetAddress
import java.net.URI
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue

class TimetableApiClient(
    private val transport: HttpExchange,
    baseUri: URI
) {
    private val base: URI = validateBaseUri(baseUri)

    suspend fun fetch(requiredGroupIds: Iterable<String>): TimetableApiSnapshot {
        val ids = sortedSetOf<String>()
        for (id in requiredGroupIds) {
            TimetableApiJson.require(TimetableApiJson.validId(id))
            ids.add(id)
            TimetableApiJson.require(ids.size <= 5000)
        }
        try {
            return withTimeout(30_000) {
                var attempt = 0
                while (true) {
                    val catalog = TimetableApiJson.catalog(get("api/v1/groups", false))
                    val byId = catalog.groups.associateBy { it.id }
                    if (ids.any { it !in byId }) throw TimetableApiException(TimetableApiFailure.UnknownRequiredGroup)
                    try {
                        val downloaded = download(catalog, ids.map { byId.getValue(it) })
                        return@withTimeout catalog.copy(downloaded = downloaded)
                    } catch (e: TimetableApiException) {
                        if (e.failure == TimetableApiFailure.SnapshotUnavailable && attempt == 0) {
                            attempt++
                            continue
                        }
                        throw e
                    }
                }
                error("unreachable")
            }
        } catch (e: kotlinx.coroutines.TimeoutCancellationException) {
            throw TimetableApiException(TimetableApiFailure.Timeout)
        } catch (e: CancellationException) {
            throw e
        } catch (e: TimetableApiException) {
            throw e
        } catch (_: JsonFail) {
            throw TimetableApiException(TimetableApiFailure.InvalidPayload)
        }
    }

    private suspend fun download(
        catalog: TimetableApiSnapshot,
        groups: List<TimetableApiGroup>
    ): Map<String, TimetableApiDownloadedGroup> = coroutineScope {
        if (groups.isEmpty()) return@coroutineScope emptyMap()
        val slots = Semaphore(4)
        val errors = ConcurrentLinkedQueue<TimetableApiException>()
        val rows = ConcurrentHashMap<String, TimetableApiDownloadedGroup>()
        val jobs = groups.map { group ->
            async {
                slots.withPermit {
                    try {
                        val escaped = escapeId(group.id)
                        val path = "api/v1/groups/$escaped/timetable?snapshotId=${catalog.meta.snapshotId}"
                        val json = get(path, true)
                        rows[group.id] = TimetableApiJson.timetable(json, catalog, group)
                    } catch (e: TimetableApiException) {
                        errors.add(e)
                        throw e
                    }
                }
            }
        }
        try {
            jobs.awaitAll()
        } catch (_: Exception) {
            val first = errors.firstOrNull() ?: throw TimetableApiException(TimetableApiFailure.Transport)
            throw errors.firstOrNull { it.failure != TimetableApiFailure.SnapshotUnavailable } ?: first
        }
        HashMap(rows)
    }

    private suspend fun get(path: String, pinned: Boolean): JsonValue {
        val url = base.toString() + path
        val reply = try {
            transport.exchange(HttpCall("GET", url, mapOf("Accept" to "application/json"), maxBytes = MAX_BYTES))
        } catch (e: CancellationException) {
            throw e
        } catch (_: HttpBodyTooLargeException) {
            throw TimetableApiException(TimetableApiFailure.BodyTooLarge)
        } catch (_: java.io.IOException) {
            throw TimetableApiException(TimetableApiFailure.Transport)
        }
        if (reply.headers.keys.any { it.equals("Content-Encoding", true) && reply.headers.getValue(it).isNotBlank() }) {
            throw TimetableApiException(TimetableApiFailure.InvalidPayload)
        }
        if (reply.body.size > MAX_BYTES) throw TimetableApiException(TimetableApiFailure.BodyTooLarge)
        if (reply.status != 200) {
            if (pinned && reply.status == 404) {
                try {
                    val error = StrictJson.parse(reply.body, 32).obj()
                    if (error.int("status") == 404 && error.text("code", 64) == "snapshot_not_found") {
                        throw TimetableApiException(TimetableApiFailure.SnapshotUnavailable)
                    }
                } catch (e: TimetableApiException) {
                    throw e
                } catch (_: Exception) {
                }
            }
            throw TimetableApiException(TimetableApiFailure.ServerUnavailable)
        }
        return try {
            StrictJson.parse(reply.body, 32)
        } catch (_: JsonFail) {
            throw TimetableApiException(TimetableApiFailure.InvalidPayload)
        }
    }

    companion object {
        private const val MAX_BYTES = 16 * 1024 * 1024

        fun escapeId(id: String): String = when (id) {
            "." -> "%2E"
            ".." -> "%2E%2E"
            else -> URLEncoder.encode(id, StandardCharsets.UTF_8).replace("+", "%20")
        }

        fun validateBaseUri(uri: URI): URI {
            val loopback = uri.isAbsolute && isLoopback(uri)
            if (!uri.isAbsolute ||
                (uri.scheme != "https" && !(uri.scheme == "http" && loopback)) ||
                !uri.userInfo.isNullOrEmpty() ||
                !uri.query.isNullOrEmpty() ||
                !uri.fragment.isNullOrEmpty()
            ) {
                throw IllegalArgumentException("Некорректный адрес API.")
            }
            val raw = uri.toString().trimEnd('/') + "/"
            return URI(raw)
        }

        fun isLoopback(uri: URI): Boolean {
            val host = uri.host ?: return false
            if (host.equals("localhost", ignoreCase = true)) return true
            val stripped = host.trimStart('[').trimEnd(']')
            return try {
                if (stripped.contains(':')) {
                    InetAddress.getByName(stripped).isLoopbackAddress
                } else {
                    val parts = stripped.split('.')
                    if (parts.size != 4) return false
                    val bytes = ByteArray(4) { i ->
                        val n = parts[i].toIntOrNull() ?: return false
                        if (n !in 0..255) return false
                        n.toByte()
                    }
                    InetAddress.getByAddress(bytes).isLoopbackAddress
                }
            } catch (_: Exception) {
                false
            }
        }
    }
}

File: src/Vograph.Core/Services/TimetableApiClient.cs
using System.Collections.Concurrent;
using System.Collections.Immutable;
using System.Net;
using System.Text;
using System.Text.Json;
using Vograph.Core.Models;

namespace Vograph.Core.Services;

public sealed class TimetableApiClient : IDisposable
{
    private const int MaxBytes = 16 * 1024 * 1024;
    private readonly HttpClient _http;
    private readonly Uri _baseUri;
    private bool _ownsHttp;

    /// <summary>Injected clients must disable redirects and use controlled decompression. Caller owns them.</summary>
    public TimetableApiClient(HttpClient httpClient, Uri baseUri)
    {
        ArgumentNullException.ThrowIfNull(httpClient);
        _baseUri = ValidateBaseUri(baseUri);
        _http = httpClient;
    }

    public static TimetableApiClient CreateOwned(Uri baseUri)
    {
        var validated = ValidateBaseUri(baseUri);
        var handler = new SocketsHttpHandler
        {
            AllowAutoRedirect = false, AutomaticDecompression = DecompressionMethods.None,
            UseCookies = false, Credentials = null, MaxConnectionsPerServer = 4
        };
        var http = new HttpClient(handler) { Timeout = Timeout.InfiniteTimeSpan };
        return new TimetableApiClient(http, validated) { _ownsHttp = true };
    }

    public async Task<TimetableApiSnapshot> FetchAsync(IEnumerable<string> requiredGroupIds,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(requiredGroupIds);
        using var deadline = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        deadline.CancelAfter(TimeSpan.FromSeconds(30));
        var ct = deadline.Token;
        try
        {
            var ids = new SortedSet<string>(StringComparer.Ordinal);
            foreach (var id in requiredGroupIds)
            {
                ct.ThrowIfCancellationRequested();
                TimetableApiJson.Require(TimetableApiJson.ValidId(id));
                ids.Add(id);
                TimetableApiJson.Require(ids.Count <= 5000);
            }
            for (var attempt = 0; ; attempt++)
            {
                // Catalog failures themselves never qualify for the pinned-generation retry.
                using var json = await GetAsync("api/v1/groups", false, ct).ConfigureAwait(false);
                var catalog = TimetableApiReader.Catalog(json.RootElement, ct);
                var byId = catalog.Groups.ToDictionary(g => g.Id, StringComparer.Ordinal);
                if (ids.Any(id => !byId.ContainsKey(id)))
                    throw new TimetableApiException(TimetableApiFailure.UnknownRequiredGroup);
                try
                {
                    var downloaded = await DownloadAsync(catalog, ids.Select(id => byId[id]), ct).ConfigureAwait(false);
                    ct.ThrowIfCancellationRequested();
                    return catalog with { DownloadedGroups = downloaded };
                }
                catch (TimetableApiException e) when (e.Failure == TimetableApiFailure.SnapshotUnavailable && attempt == 0)
                {
                    ct.ThrowIfCancellationRequested();
                }
            }
        }
        catch (OperationCanceledException)
        {
            cancellationToken.ThrowIfCancellationRequested();
            throw new TimetableApiException(TimetableApiFailure.Timeout);
        }
        catch (InvalidOperationException)
        {
            // JsonElement can fail lazily on invalid escaped Unicode while mapping fields.
            throw new TimetableApiException(TimetableApiFailure.InvalidPayload);
        }
    }

    private async Task<ImmutableDictionary<string, TimetableApiDownloadedGroup>> DownloadAsync(
        TimetableApiSnapshot catalog, IEnumerable<TimetableApiGroup> groups, CancellationToken ct)
    {
        using var failed = CancellationTokenSource.CreateLinkedTokenSource(ct);
        using var slots = new SemaphoreSlim(4);
        var errors = new ConcurrentQueue<TimetableApiException>();
        var rows = new ConcurrentDictionary<string, TimetableApiDownloadedGroup>(StringComparer.Ordinal);
        var tasks = groups.Select(async group =>
        {
            await slots.WaitAsync(failed.Token).ConfigureAwait(false);
            try
            {
                var escaped = group.Id is "." or ".." ? group.Id.Replace(".", "%2E") : Uri.EscapeDataString(group.Id);
                var path = $"api/v1/groups/{escaped}/timetable?snapshotId={catalog.Meta.SnapshotId:D}";
                using var json = await GetAsync(path, true, failed.Token).ConfigureAwait(false);
                rows[group.Id] = TimetableApiReader.Timetable(json.RootElement, catalog, group, failed.Token);
            }
            catch (TimetableApiException e)
            {
                errors.Enqueue(e);
                failed.Cancel();
                throw;
            }
            finally { slots.Release(); }
        }).ToArray();
        try { await Task.WhenAll(tasks).ConfigureAwait(false); }
        catch
        {
            ct.ThrowIfCancellationRequested();
            // Never let a racing snapshot-not-found hide a malformed/unavailable response.
            if (errors.TryPeek(out var first))
                throw errors.FirstOrDefault(e => e.Failure != TimetableApiFailure.SnapshotUnavailable) ?? first;
            throw;
        }
        return rows.ToImmutableDictionary(StringComparer.Ordinal);
    }

    private async Task<JsonDocument> GetAsync(string path, bool pinned, CancellationToken ct)
    {
        try
        {
            // Base is validated/canonicalized, path is internal and IDs escaped. Preserve opaque dot IDs.
            var uri = new Uri(_baseUri.AbsoluteUri + path,
                new UriCreationOptions { DangerousDisablePathAndQueryCanonicalization = true });
            using var request = new HttpRequestMessage(HttpMethod.Get, uri);
            request.Headers.Accept.ParseAdd("application/json");
            using var response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, ct).ConfigureAwait(false);
            if (response.StatusCode != HttpStatusCode.OK)
            {
                if (pinned && response.StatusCode == HttpStatusCode.NotFound)
                {
                    using var error = await ReadJsonAsync(response.Content, 4096, ct).ConfigureAwait(false);
                    var root = TimetableApiJson.Object(error.RootElement);
                    if (TimetableApiJson.Integer(root, "status") == 404
                        && TimetableApiJson.Text(root, "code", 64) == "snapshot_not_found")
                        throw new TimetableApiException(TimetableApiFailure.SnapshotUnavailable);
                }
                throw new TimetableApiException(TimetableApiFailure.ServerUnavailable);
            }
            return await ReadJsonAsync(response.Content, MaxBytes, ct).ConfigureAwait(false);
        }
        catch (JsonException) { throw new TimetableApiException(TimetableApiFailure.InvalidPayload); }
        catch (DecoderFallbackException) { throw new TimetableApiException(TimetableApiFailure.InvalidPayload); }
        catch (HttpRequestException) { throw new TimetableApiException(TimetableApiFailure.Transport); }
        catch (IOException) { throw new TimetableApiException(TimetableApiFailure.Transport); }
    }

    private static async Task<JsonDocument> ReadJsonAsync(HttpContent content, int limit, CancellationToken ct)
    {
        if (content.Headers.ContentLength > limit) throw new TimetableApiException(TimetableApiFailure.BodyTooLarge);
        if (content.Headers.ContentEncoding.Count != 0) throw new TimetableApiException(TimetableApiFailure.InvalidPayload);
        using var stream = await content.ReadAsStreamAsync(ct).ConfigureAwait(false);
        using var bytes = new MemoryStream();
        var buffer = new byte[16384];
        while (true)
        {
            ct.ThrowIfCancellationRequested();
            var read = await stream.ReadAsync(buffer.AsMemory(0, Math.Min(buffer.Length, limit - (int)bytes.Length + 1)), ct)
                .ConfigureAwait(false);
            if (read == 0) break;
            if (bytes.Length + read > limit) throw new TimetableApiException(TimetableApiFailure.BodyTooLarge);
            bytes.Write(buffer, 0, read);
        }
        ct.ThrowIfCancellationRequested();
        _ = new UTF8Encoding(false, true).GetCharCount(bytes.GetBuffer(), 0, (int)bytes.Length);
        return JsonDocument.Parse(bytes.GetBuffer().AsMemory(0, (int)bytes.Length), new JsonDocumentOptions { MaxDepth = 32 });
    }

    private static Uri ValidateBaseUri(Uri uri)
    {
        ArgumentNullException.ThrowIfNull(uri);
        var loopback = uri.IsAbsoluteUri && (uri.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase)
            || (IPAddress.TryParse(uri.Host.Trim('[', ']'), out var address) && IPAddress.IsLoopback(address)));
        if (!uri.IsAbsoluteUri || (uri.Scheme != "https" && !(uri.Scheme == "http" && loopback))
            || uri.UserInfo.Length != 0 || uri.Query.Length != 0 || uri.Fragment.Length != 0)
            throw new ArgumentException("Некорректный адрес API.", nameof(uri));
        return new Uri(uri.AbsoluteUri.TrimEnd('/') + "/");
    }

    public void Dispose()
    {
        if (_ownsHttp) _http.Dispose();
    }
}
